### ISSUE TITLE
CORS Implementation 

### ISSUE THREAD

**Issue by tsirolnik** [tmpvar/jsdom#2024]
Simply put, as it seems, the current version of jsdom doesn't follow the [CORS guidelines](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) of preflighting and etc. A thing that results in blockage of requests instead of preflighting.

This is more of a feature request than an issue. I'd be happy to try and implement it

Also, correct me if I'm wrong, but jsdom stops executing once a CORS error appears?

Thanks!

**Comment by domenic** [comment]
We have quite a lot of a CORS implementation, although it is not perfect. We pass a lot of the web platform tests for CORS; the ones we fail are listed in https://github.com/tmpvar/jsdom/blob/master/test/web-platform-tests/to-run.yaml . It does appear some of the preflight stuff is failing, especially when redirects are involved. But some preflights are done, e.g. we pass https://github.com/w3c/web-platform-tests/blob/master/cors/preflight-failure.htm .

Help fixing our implementation to pass more of those tests would definitely be appreciated.

jsdom should still execute when a CORS error appears. A test case (following the issue template, i.e. minimal and with code that works in browsers) would be helpful in that regard.

**Comment by tsirolnik** [comment]
Thanks for the quick feedback. 

> It does appear some of the preflight stuff is failing, especially when redirects are involved

This is exactly the case I've found that seems to fail

> Help fixing our implementation to pass more of those tests would definitely be appreciated.

I'll try to implement it and write a test for it.

>jsdom should still execute when a CORS error appears. A test case (following the issue template, i.e. minimal and with code that works in browsers) would be helpful in that regard.

Does the jsdom error that comes up blocks further executing? As in - If a CORS error is detected, does it stop executing after throwing the jsdom error? I'm talking about the thing that happens in `xmlhttprequest.js` 

```js
  function validCORSHeaders(xhr, response, flag, properties, origin) {
    const acaoStr = response.headers["access-control-allow-origin"];
    const acao = acaoStr ? acaoStr.trim() : null;
    if (acao !== "*" && acao !== origin) {
      properties.error = "Cross origin " + origin + " forbidden";
      dispatchError(xhr);
      return false;
    }

```

Does dispatch error causes jsdom to stop running or does it continue in the execution?

**Comment by ScottAlbertine** [comment]
I am also running into another issue in that function, due to jsdom not following that same cors standard.  Jsdom expects all cross-origin responses to have access-control-allow-headers on them, whereas the cors standard (and my server's standard implementation) only has them on preflight (OPTIONS) responses.  This means that I cannot make a HTTP BasicAuth login call to my server in a test, as the Authorization header must exist on the non-preflight request, and jsdom disallows headers that aren't in the access-control-allow-headers header in the response to the non-preflight request, and my server doesn't send the access-control-allow-headers header in response to non-preflight requests, as is the cors standard.

**Comment by Sebmaster** [comment]
> Does dispatch error causes jsdom to stop running or does it continue in the execution?

It continues to run. It should dispatch an `error` event to your `window` - and, if uncaught, then forward that to the virtual console (as `jsdomError`, I think?).

**Comment by ScottAlbertine** [comment]
@Sebmaster @domenic It does not continue to run, it abandons the parsing of the response on line 841 of xmlhttprequest.js, since the validCorsHeader function returns false after dispatching the error.  Here's a link to the actual return statement that stops it running: https://github.com/tmpvar/jsdom/blob/0db4270773d4f4b7f50bd66cdb776f7811670fd7/lib/jsdom/living/xmlhttprequest.js#L841

**Comment by Sebmaster** [comment]
Well, the CORS response will obviously be blocked if it's not allowed, but that shouldn't impact any other parts of jsdom.

**Comment by ScottAlbertine** [comment]
That's not in the code that handles just preflight requests, that's in the code that handles *all* cross-origin requests, even GET requests.  It means I can't make *any* cross-origin xhr calls in a test that's using jsdom, without literally re-writing my server's CORS handling to no longer follow the actual CORS spec.

**Comment by domenic** [comment]
> I'll try to implement it and write a test for it.

This should not be necessary; just enable the existing tests. See CONTRIBUTING.md.

> Does the jsdom error that comes up blocks further executing?

No. It's just like a browser.

> It means I can't make any cross-origin xhr calls in a test that's using jsdom, without literally re-writing my server's CORS handling to no longer follow the actual CORS spec.

Patches (with newly-enabled tests) welcome.

**Comment by ScottAlbertine** [comment]
So, I've spent a day on this, and I've found 2 problems.  The first is the one I mentioned above, where the lack of access-control-allow-headers on the regular (the POST after a preflight) response causes an error to be thrown.  This should be an **immediate** red flag for y'all, because it demonstrates that, despite the access-control-allow-headers field not containing a header that's on the regular request, the regular request **still gets sent**.  We know this because its response not only gets parsed, but it's actively used to throw an error.  Not only that, it turns out the disallowed headers aren't even removed from the regular request!  This means that your CORS system is not only broken, it's so broken that it doesn't actually prevent me from sending cross-origin requests with malicious headers, **and** it's so broken that it actively covers up that fact by throwing the error based on the regular response headers, which is why your tests didn't pick up on this before.

The key is that xmlhttprequest.js:841 line that I mentioned in my comment above.  It's only called for regular requests, **not** for preflight requests.  It also contains a check for access-control-allow-headers, which should only be checked on preflight responses.  However, the code path leading to that spot is so convoluted that it took me most of today to figure out where to actually check the headers for the preflight request.  It's in a completely different file, xhr-utils, between lines 284 and 285.  I've added a new validCORSPreflightHeaders check there, and kept a smaller version of it on line 841 to handle non-preflighted GET requests.  To do this, however, I've had to move most of the private xmlhttprequest methods into xhr-utils, and convert that module into one that returns a constructor that takes a window as a parameter.  This is so I can call new Event() in case the CORS headers are invalid in the xhr-utils check of them.

While I was in there, I also fixed the list of headers actually sent on preflight requests themselves too, to be in line with the CORS spec, which was also a problem.

Here is my pull request: https://github.com/tmpvar/jsdom/pull/2027

Please let me know what branch to add it to so it can get backported into the 9.x series of jsdom that Jest uses, as that's where I actually need it to be.  I'll follow up with Jest once I've got it there.

**Comment by domenic** [comment]
We don't do backports, I'm sorry to say, but thank you for the pull request!

**Comment by ScottAlbertine** [comment]
...what do you mean you don't do backports?  You're a hard dependency of **Jest**.  This means I am unable to **log in** to my server in any Jest test, it's blocking me from writing literally any integration tests that require login functionality, which is 90% of them.  Am I going to have to wait for Jest to update to jsdom 11.4.0 whenever that comes out?  Is there another method to fix this issue so that I can actually continue with my day job, other than literally commenting out the problematic code in my node_modules directory?

**Comment by domenic** [comment]
We don't maintain old versions of our software, as this is a spare-time project and we don't enjoy the maintenance burden, but instead prefer to make the best version of jsdom (the latest one) even better.

You're welcome to investigate alternatives.

**Comment by ScottAlbertine** [comment]
Oh.  I thought y'all were owned by Facebook, seeing as they own Jest and it depends directly on jsdom.  You may want to seriously consider contacting Facebook (who maintains Jest) and asking them to assign a developer to this project to bring it up to speed, since it's required for Jest tests to function properly.

**Comment by Zirro** [comment]
@ScottAlbertine I think [this](https://github.com/facebook/jest/pull/4720) jest issue might be of interest to you.

**Comment by ScottAlbertine** [comment]
Oh, April then... cool, yeah, I'm ok with commenting out the problematic code in my node_modules folder until Node 4 reaches EOL and Facebook can update Jest to use jsdom 11.x.  For anyone who finds this thread and is having similar issues, the problematic code for me is lines 1019 through 1029 of node_modules/jsdom/lib/jsdom/living/xmlhttprequest.js as of jsdom 9.12.0, which is the official version used by Jest at the time of this writing.

**Comment by Zirro** [comment]
@ScottAlbertine Looks like you're in luck. Yesterday they took [the decision](https://github.com/facebook/jest/pull/4767#discussion_r147131131) to drop Node 4 for the next major version, and [upgrade](https://github.com/facebook/jest/pull/4770) to jsdom 11 in the process.

**Comment by SimenB** [comment]
@ScottAlbertine You were never blocked by Jest upgrading to jsdom@9, you can use https://github.com/Micromeritics/jest-environment-jsdom-11.0.0 in current versions of Jest

**Comment by domenic** [comment]
I guess this was fixed in https://github.com/jsdom/jsdom/pull/2027, so let me close this.
